### PR TITLE
Patches: Adds db_nodes_autoinc patch

### PR DESCRIPTION
**Pull request by tomponline**
Recreates nodes table id column with AUTOINCREMENT.

Related to https://github.com/canonical/raft/pull/176
Related to https://github.com/lxc/lxd/pull/8521

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>

### CODE REVIEW

File: lxd/patches.go [hunk 1]
```diff
@@ -181,6 +182,51 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+func patchDBNodesAutoInc(name string, d *Daemon) error {
+	row := d.State().Cluster.DB().QueryRow("SELECT sql FROM sqlite_master WHERE name = 'nodes'")
+
+	var schemaSQL string
+	err := row.Scan(&schemaSQL)
+	if err != nil {
+		return err
+	}
+
+	if strings.Contains(schemaSQL, "id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL") {
```
Comment (tomponline) [inline lxd/patches.go:196]: @stgraber not sure if this is what you had in mind to only run on leader, but this should prevent it running multiple times.

File: lxd/patches.go [hunk 2]
```diff
@@ -181,6 +182,51 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+		logger.Debugf("Skipping %q patch as node table id column already AUTOINCREMENT", name)
+		return nil // Nothing to do.
+	}
+
+	_, err = d.State().Cluster.DB().Exec(`
+PRAGMA foreign_keys = OFF;
+PRAGMA legacy_alter_table = ON;
```
Comment (stgraber) [inline lxd/patches.go:203]: Isn't that global? If so, you definitely need to revert that at the end.

File: lxd/patches.go (hunk 1 above)
Comment (stgraber) [reply lxd/patches.go:196]: That's a good protection to have.

My thought was to have call `node.ClusterAddress(d.db)` combined with `d.gateway.LeaderAddress()` and if they're not equal, then skip as we're not on the leader.

File: lxd/patches.go (hunk 2 above)
Comment (tomponline) [reply lxd/patches.go:203]: Mmm yeah if the connection is held open after patches then yes. I'll also add a check for leader so there's no chance of a race between checking the existing schema and applying the fix if missing.

**Comment by tomponline** [comment]
@stgraber I've updated this with the changes suggested, and have modified the way that the patch works to loop waiting for the patch to be applied on the leader (if clustered), because in my testing on a cluster, I found that if there was an error applying the patch (in my case I'd made a typo in the SQL) the leadership seemed to flip flop and eventually allowed the patch to be skipped and considered "applied" on all nodes due to the fact that at some point each of the nodes run the patch when not a leader and skipped apply it.

Now I've updated the patch so that each node checks if the schema needs the patch, and if so, then only applies it if its the leader, and if not waits for 1s and then checks the schema again, and repeats. 

Now in my test I see this:

Node 1:

```
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Initializing and checking storage pool \"local\"" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount started" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount finished" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=info msg="Applying patch \"db_nodes_autoinc\"" 
t=2021-03-05T10:13:19+0000 lvl=warn msg="Waiting for \"db_nodes_autoinc\" patch to be applied on leader cluster member" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:20+0000 lvl=dbug msg="Skipping \"db_nodes_autoinc\" patch as \"nodes\" table id column already AUTOINCREMENT" 
t=2021-03-05T10:13:20+0000 lvl=info msg="Initializing daemon storage mounts" 
```

Node 2:

```
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Initializing and checking storage pool \"local\"" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount started" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount finished" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=info msg="Applying patch \"db_nodes_autoinc\"" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:20+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:20+0000 lvl=info msg="Initializing daemon storage mounts" 
```

Node 3:

```
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Initializing and checking storage pool \"local\"" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount started" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount finished" driver=dir pool=local
t=2021-03-05T10:13:20+0000 lvl=info msg="Applying patch \"db_nodes_autoinc\"" 
t=2021-03-05T10:13:20+0000 lvl=warn msg="Waiting for \"db_nodes_autoinc\" patch to be applied on leader cluster member" 
t=2021-03-05T10:13:21+0000 lvl=dbug msg="Skipping \"db_nodes_autoinc\" patch as \"nodes\" table id column already AUTOINCREMENT" 
t=2021-03-05T10:13:21+0000 lvl=info msg="Initializing daemon storage mounts" 
```

So it applied on node 2, and then the other two noticed and proceeded.

**Comment by tomponline** [comment]
jenkins: test this please

**Pull request merged**